### ISSUE TITLE
Theme improvements: override default global song/bible themes with custom item theme

### ISSUE THREAD

**Issue by DeeTigIT** [quelea-projection/Quelea#44]
**Here's a testcase on the default global song/bible themes in combination with the 'Allow custom item themes to override global theme' user option.**

A custom bible theme can override the default bible theme.
But a custom song theme isn't shown when the override option is set.
**The details can be found in the docx file.**

[theme improvements (Allow custom item themes to override global theme) - test results.docx](https://github.com/quelea-projection/Quelea/files/2425415/theme.improvements.Allow.custom.item.themes.to.override.global.theme.-.test.results.docx)

**Comment by berry120** [comment]
Hmm, afraid I can't seem to reproduce this - does this happen consistently, or only with an open / close of schedule / some other type of behaviour?

**Comment by DeeTigIT** [comment]
It's reproducable and even with the schedule build up from ground again. There's a peculiar thing in the schedule with the song item equipped with an image custom theme (sunset sky.jpg), I noticed. Although the picture is projected (use the default Quelea global song/bible themes; no global theme overrides), it doesn't show up when re-inspecting the theme properties of the item **after saving and re-opening the schedule.** (during the inital schedule build it does show up)
Items with a colour or video custom theme don't have this: the colour of video does show up when re-inspecting the theme properties of the item.
I added the schedule (plz. rename theme .qsch.zip to .qsch).

![s1](https://user-images.githubusercontent.com/39883796/46186748-70755280-c2e0-11e8-9acc-46391fb6593e.JPG)

![s2](https://user-images.githubusercontent.com/39883796/46186755-7703ca00-c2e0-11e8-9072-221252f59ab9.JPG)

![s3](https://user-images.githubusercontent.com/39883796/46186758-7bc87e00-c2e0-11e8-853d-6e04988c14d7.JPG)

[theme improvements.qsch.zip](https://github.com/quelea-projection/Quelea/files/2426659/theme.improvements.qsch.zip)

**Comment by DeeTigIT** [comment]
I've just opened a new issue:
Theme improvements: show previously set song theme properties in a re-opened schedule #56

For the above described observation during testing of these #44 changes:
_There's a peculiar thing in the schedule with the song item equipped with an image custom theme (sunset sky.jpg), I noticed. Although the picture is projected (use the default Quelea global song/bible themes; no global theme overrides), it doesn't show up when re-inspecting the theme properties of the item after saving and re-opening the schedule. (during the inital schedule build it does show up)_

Turns out it's also present in the 2018.1 stable release (and thus not related to these #44 changes).

**Comment by berry120** [comment]
Managed to recreate this now 👍 I've attempted a fix, could you check? Thanks!

**Comment by DeeTigIT** [comment]
The fix seems to work only when building a new schedule: Step 5 - Projection then shows the expected custom item theme image "sunset sky.jpg" instead of the global "sunset lake.jpg". That's ok now.
But ... when the schedule is saved and re-opened in the same Q session, the override doesn't occur. Closing Q and opening the schedule has the same effect: no override.

**Comment by DeeTigIT** [comment]
Same conclusion:
_But ... when the schedule is saved and re-opened in the same Q session, the override doesn't occur. Closing Q and opening the schedule has the same effect: no override._

I also noticed 2 regressions:
- When an additional song is copied to the schedule (without setting a custom theme on it), then the global song theme isn't used. It seems the default Q song theme is used instead. 
- When different global song and bible themes are use, the global bible theme is used for songs. (probably this one was also present in attempt number 1).

To sharpen my understanding of the Q themes, I made a document with expected themes depending on the override property, the global song/bible themes and the custom song/bible themes.
**Could you please verify this diagram?** This could clear-up any misunderstandings on my side. 

[expected_themes.xlsx](https://github.com/quelea-projection/Quelea/files/2546403/expected_themes.xlsx)

**Comment by berry120** [comment]
Yes, I think your understanding is correct from that spreadsheet - I just need to have another go at reproducing this and getting the correct behaviour sorted out.

**Comment by DeeTigIT** [comment]
This works great now! You can close the call. 

Next sunday, we will use this theme override feature in our service. And coming thursday evening, I'll introduce this to the other members of our church projection team in an introductory Quelea course.

**Comment by berry120** [comment]
Awesome, glad we've finally got there. Thanks very much for raising the issue!

**Comment by DeeTigIT** [comment]
Hello Michael,

In preparing the Quelea mini course for tomorrow, I encountered this: 

a)
Initial settings: both a {Q global song theme} and a {Q global bible theme) are set (they differ by the way), theme_override_property_set

b)
During the build of a new schedule (when songs and bible items not yet have custom themes) , the {Q global song **default** theme} is applied unexpectedly, instead of the user defined {Q global song theme}.
For bible items the user defined {Q global bible theme) is applied, as expected. So, for bible items no problems.

c)
But here it comes: after saving and re-opening the schedule: for song items the user defined {Q global song theme} is applied as well.
So, saving and re-opening the schedule corrects the above issue for song items.


Yesterday, I tested with an existing schedule from the past. (no problem encountered)
Today with a new schedule on a different computer.
My apologies for this omission in testing.

**Is this fixable, somehow?** 

(so that song items in an unsaved, new schedule are already shown with the user defined  {Q global song theme})

Regards and blessings,

Dirk

**Comment by DeeTigIT** [comment]
A song with custom theme set (your test) behaves all right : the custom song theme is indeed shown.

But for a song without custom theme (my test) the global song theme isn't used during the build of a new, unsaved schedule. Only after saving and re-opening the global theme is used for such songs.

In the data beneath:
- div 85 : custom song theme YES
- div 86 : custom song theme NO

**New, unsaved sched:**

![s2](https://user-images.githubusercontent.com/39883796/50362231-3a561080-0567-11e9-89cf-c7204424b95c.JPG)

![s3](https://user-images.githubusercontent.com/39883796/50362238-3de99780-0567-11e9-834e-0fbb6099f20b.JPG)

**Save and re-open sched:**

![s4](https://user-images.githubusercontent.com/39883796/50362260-5063d100-0567-11e9-96e4-0de0f3a7d4e3.JPG)

![s5](https://user-images.githubusercontent.com/39883796/50362263-548fee80-0567-11e9-9813-3aac9b2525b0.JPG)

Global themes:

![s1](https://user-images.githubusercontent.com/39883796/50362206-2ad6c780-0567-11e9-9664-cc91607b8a0d.JPG)

**Comment by berry120** [comment]
Afraid I'm still struggling to reproduce - I've set the "theme override" option, set a default song theme, and added a song *without* a custom theme to the schedule - for me at least, this displays the default song theme as intended.

Don't suppose there's any other options set that may help in trying to work out what's going on?

**Comment by DeeTigIT** [comment]
Thanks for letting me know this. I’ll will have a closer look.  There is a challenge here.

Van: Michael Berry [mailto:notifications@github.com]
Verzonden: zaterdag 22 december 2018 12:39
Aan: quelea-projection/Quelea
CC: DeeTigIT; Author
Onderwerp: Re: [quelea-projection/Quelea] Theme improvements: override default global song/bible themes with custom item theme (#44)


Afraid I'm still struggling to reproduce - I've set the "theme override" option, set a default song theme, and added a song without a custom theme to the schedule - for me at least, this displays the default song theme as intended.

Don't suppose there's any other options set that may help in trying to work out what's going on?

**Comment by berry120** [comment]
This should now be fixed - @ArvidNy managed to reproduce! Could you check again when convenient? Thanks!

**Comment by DeeTigIT** [comment]
Hi Michael,

I did the test again with the latest CI-releases in 2 configs:


-          quelea-CI-UNSTABLE-x64-windows-install.exe (implicit use of JAVA version 1.8.0_111).

-          quelea-CI-UNSTABLE-crossplatform-install.jar + jre-10.0.2_windows-x64_bin.exe

Seems there’s a status quo here as  I’m getting the same test results.
Is the fix already incorporated in the CI downloads?

Regards,

Dirk


Van: Michael Berry [mailto:notifications@github.com]
Verzonden: dinsdag 8 januari 2019 18:52
Aan: quelea-projection/Quelea
CC: DeeTigIT; Author
Onderwerp: Re: [quelea-projection/Quelea] Theme improvements: override default global song/bible themes with custom item theme (#44)


This should now be fixed - @ArvidNy<https://github.com/ArvidNy> managed to reproduce! Could you check again when convenient? Thanks!

**Comment by ArvidNy** [comment]
Hi Dirk,

Seems Michael forgot a variable in the changes he made yesterday but I just added the missing line in, so once it the CI release is built you should be able to give it a new attempt.

Thanks,

Arvid

**Comment by DeeTigIT** [comment]
Thank you for the quick response on this. I will check for a new CI tomorrow.

Verzonden vanaf mijn Windows 10-apparaat

________________________________
Van: Michael Berry <notifications@github.com>
Verzonden: Wednesday, January 9, 2019 10:39:43 PM
Aan: quelea-projection/Quelea
CC: DeeTigIT; Author
Onderwerp: Re: [quelea-projection/Quelea] Theme improvements: override default global song/bible themes with custom item theme (#44)


@ArvidNy<https://github.com/ArvidNy> Oops. Sorry about that. Thanks for fixing!

**Comment by ArvidNy** [comment]
Great, thanks for letting us know!
